external/fv3fit/fv3fit/pytorch/cyclegan/cyclegan_trainer.py

```diff
@@ -0,0 +1,424 @@
+    """
+
+    generator_optimizer: OptimizerConfig = dataclasses.field(
+        default_factory=lambda: OptimizerConfig("Adam")
+    )
+    discriminator_optimizer: OptimizerConfig = dataclasses.field(
+        default_factory=lambda: OptimizerConfig("Adam")
+    )
+    generator: "GeneratorConfig" = dataclasses.field(
+        default_factory=lambda: GeneratorConfig()
+    )
+    discriminator: "DiscriminatorConfig" = dataclasses.field(
+        default_factory=lambda: DiscriminatorConfig()
+    )
+    identity_loss: LossConfig = dataclasses.field(default_factory=LossConfig)
+    cycle_loss: LossConfig = dataclasses.field(default_factory=LossConfig)
+    gan_loss: LossConfig = dataclasses.field(default_factory=LossConfig)
+    identity_weight: float = 0.5
+    cycle_weight: float = 1.0
+    generator_weight: float = 1.0
+    discriminator_weight: float = 1.0
+
+    def build(
+        self, n_state: int, n_batch: int, state_variables, scalers
+    ) -> "CycleGANTrainer":
+        generator_a_to_b = self.generator.build(n_state)
+        generator_b_to_a = self.generator.build(n_state)
+        discriminator_a = self.discriminator.build(n_state)
+        discriminator_b = self.discriminator.build(n_state)
+        optimizer_generator = self.generator_optimizer.instance(
+            itertools.chain(
+                generator_a_to_b.parameters(), generator_b_to_a.parameters()
+            )
+        )
+        optimizer_discriminator = self.discriminator_optimizer.instance(
+            itertools.chain(discriminator_a.parameters(), discriminator_b.parameters())
+        )
+        return CycleGANTrainer(
+            cycle_gan=CycleGAN(
+                model=CycleGANModule(
+                    generator_a_to_b=generator_a_to_b,
+                    generator_b_to_a=generator_b_to_a,
+                    discriminator_a=discriminator_a,
+                    discriminator_b=discriminator_b,
+                ).to(DEVICE),
+                state_variables=state_variables,
+                scalers=_merge_scaler_mappings(scalers),
+            ),
+            optimizer_generator=optimizer_generator,
+            optimizer_discriminator=optimizer_discriminator,
+            identity_loss=self.identity_loss.instance,
+            cycle_loss=self.cycle_loss.instance,
+            gan_loss=self.gan_loss.instance,
+            batch_size=n_batch,
+            identity_weight=self.identity_weight,
+            cycle_weight=self.cycle_weight,
+            generator_weight=self.generator_weight,
+            discriminator_weight=self.discriminator_weight,
+        )
+
+
+def _merge_scaler_mappings(
+    scaler_tuple: Tuple[Mapping[str, StandardScaler], Mapping[str, StandardScaler]]
+) -> Mapping[str, StandardScaler]:
+    scalers = {}
+    for prefix, scaler_map in zip(("a_", "b_"), scaler_tuple):
+        for key, scaler in scaler_map.items():
+            scalers[prefix + key] = scaler
+    return scalers
+
+
+class ReplayBuffer:
```

Comment:
Did you try training without this, ie, do we really need it? If so it seems like it'd be nice to have a generic random buffer class that takes in a sequence of samples and returns a random subset of the buffer, without the bespoke types and methods here that make this code unlikely to be reused anywhere else.

Reply:
This code is pulled directly from the reference implementation of the CycleGAN. They reported in their paper that they get oscillations in model behavior during training without it (which is referenced in the comment below).

I don't know that we have a need for this anywhere other than the CycleGAN training, `.shuffle()` mostly does the job. I'd prefer to keep it private here. I can rename it to be module-private with `_`.

Reply:
I'll switch out the implementation to the one used in the [reference implementation of the author](https://github.com/junyanz/pytorch-CycleGAN-and-pix2pix/blob/eb6ae80412e23c09b4317b04d889f1af27526d2d/util/image_pool.py), with a url reference.

Reply:
Done.